### PR TITLE
Tcp filter report delta

### PR DESCRIPTION
**Pull request by JimmyCYJ**
**What this PR does / why we need it**: Let MixerClient send periodical reports for long TCP connections. Use config TcpClientConfig::report_interval to set report interval. If report_interval is not set, or report_interval is not correct, then use 10 seconds as default interval.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #[539](https://github.com/istio/proxy/issues/539)

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: control/include/tcp/controller.h [hunk 1]
```diff
@@ -54,6 +54,10 @@ class Controller {
 
   // Get statistics.
   virtual void GetStatistics(::istio::mixer_client::Statistics* stat) const = 0;
+
+  virtual const ::istio::mixer_client::Environment& environment() const = 0;
+
+  virtual std::chrono::milliseconds report_interval_ms() const = 0;
```
Comment (qiwzhang) [inline control/include/tcp/controller.h]: Not need to expose these two.   It is not this lib job to keep track of config.  
* proxy can use Envoy timer directly.    
* proxy can make a copy of report_interval.

File: control/include/tcp/request_handler.h [hunk 2]
```diff
@@ -38,6 +38,11 @@ class RequestHandler {
   // Make report call.
   // This can be called multiple times for long connection.
```
Comment (qiwzhang) [inline control/include/tcp/request_handler.h]: Add TODO for this function to remove it.

File: control/src/tcp/attributes_builder.cc [hunk 3]
```diff
@@ -41,23 +41,37 @@ void AttributesBuilder::ExtractCheckAttributes(CheckData* check_data) {
...
+  if (is_final_report) {
+    // TODO(JimmyCYJ): After Envoy TCP filter switches to use
+    // Report(ReportData* report_data, bool is_final_report) call, do not
+    // report "connection.received.bytes" and "connection.sent.bytes" in final
+    // report.
+    builder.AddInt64(AttributeName::kConnectionReceviedBytes,
+                     info.received_bytes);
+    builder.AddInt64(AttributeName::kConnectionReceviedTotalBytes,
+                     info.received_bytes);
+    builder.AddInt64(AttributeName::kConnectionSendBytes, info.send_bytes);
```
Comment (qiwzhang) [inline control/src/tcp/attributes_builder.cc]: need to minus last send_bytes too

File: control/src/tcp/attributes_builder.h [hunk 4]
```diff
@@ -27,16 +27,24 @@ namespace tcp {
...
 
  private:
   // The request context object.
   RequestContext* request_;
+
+  // Records reported information in last Report() call. This is needed for
+  // calculating delta information which will be sent in periodical report.
+  // Delta information includes incremented sent bytes and received bytes
+  // between last report and this report.
+  ReportData::ReportInfo last_report_info_;
```
Comment (qiwzhang) [inline control/src/tcp/attributes_builder.h]: I prefer not to use AttributesBuilder to store "last" info.  "last" should be stored in request_impl.

AttributeBuilder is one shot think, as it always allocated from stack as

Builder b(&attributes)
b.Setxxxx

**Comment by JimmyCYJ** [comment]
Please take a look.

File: control/src/tcp/request_handler_impl.h [hunk 5]
```diff
@@ -34,14 +35,30 @@ class RequestHandlerImpl : public RequestHandler {
...
+
  private:
   // The request context object.
   RequestContext request_context_;
 
   // The client context object.
   std::shared_ptr<ClientContext> client_context_;
+
+  // Builds report attributes.
+  AttributesBuilder report_attributes_builder_;
```
Comment (qiwzhang) [inline control/src/tcp/request_handler_impl.h]: Not need to store Builder

File: control/src/tcp/request_handler_impl.h [hunk 6]
```diff
@@ -34,14 +35,30 @@ class RequestHandlerImpl : public RequestHandler {
...
+
+  // Records reported information in last Report() call. This is needed for
+  // calculating delta information which will be sent in periodical report.
+  // Delta information includes incremented sent bytes and received bytes
+  // between last report and this report.
+  ReportData::ReportInfo last_report_info_;
```
Comment (qiwzhang) [inline control/src/tcp/request_handler_impl.h]: Are you sure all "last" fields reset to 0

File: control/src/tcp/attributes_builder.cc [hunk 7]
```diff
@@ -41,23 +41,35 @@ void AttributesBuilder::ExtractCheckAttributes(CheckData* check_data) {
...
-    builder.AddString(AttributeName::kCheckErrorMessage,
-                      request_->check_status.ToString());
+  if (is_final_report) {
+    builder.AddInt64(AttributeName::kConnectionReceviedBytes,
+                     info.received_bytes - last_report_info->received_bytes);
+    builder.AddInt64(AttributeName::kConnectionReceviedTotalBytes,
+                     info.received_bytes);
+    builder.AddInt64(AttributeName::kConnectionSendBytes,
+                     info.send_bytes - last_report_info->send_bytes);
+    builder.AddInt64(AttributeName::kConnectionSendTotalBytes, info.send_bytes);
```
Comment (qiwzhang) [inline control/src/tcp/attributes_builder.cc]: above 4 attributes should be handled the same way regardless is_final or not.

**Comment by JimmyCYJ** [comment]
Please take a look.

**Pull request merged**